src/Microsoft.ML.DataView/VectorType.cs

```diff
@@ -57,33 +57,20 @@ public VectorType(PrimitiveDataViewType itemType, params int[] dimensions)
         }
 
         /// <summary>
-        /// Creates a <see cref="VectorType"/> whose dimensionality information is the given <paramref name="template"/>'s information.
-        /// </summary>
-        [BestFriend]
-        internal VectorType(PrimitiveDataViewType itemType, VectorType template)
-            : base(GetRawType(itemType))
-        {
-            Contracts.CheckValue(template, nameof(template));
-
-            ItemType = itemType;
-            Dimensions = template.Dimensions;
-            Size = template.Size;
-        }
-
-        /// <summary>
-        /// Creates a <see cref="VectorType"/> whose dimensionality information is the given <paramref name="template"/>'s information,
-        /// concatenated with the specified <paramref name="dims"/>.
+        /// Constructs a potentially multi-dimensional vector type.
         /// </summary>
-        [BestFriend]
-        internal VectorType(PrimitiveDataViewType itemType, VectorType template, params int[] dims)
+        /// <param name="itemType">The type of the items contained in the vector.</param>
+        /// <param name="dimensions">The dimensions. Note that, like <see cref="Dimensions"/>, must be non-empty, with all
+        /// non-negative values. Also, because <see cref="Size"/> is the product of <see cref="Dimensions"/>, the result of
+        /// multiplying all these values together must not overflow <see cref="int"/>.</param>
+        public VectorType(PrimitiveDataViewType itemType, ImmutableArray<int> dimensions)
             : base(GetRawType(itemType))
         {
-            Contracts.CheckValue(template, nameof(template));
-            Contracts.CheckParam(Utils.Size(dims) > 0, nameof(dims));
-            Contracts.CheckParam(dims.All(d => d >= 0), nameof(dims));
+            Contracts.CheckParam(dimensions.Length > 0, nameof(dimensions));
```

Comment:
>dimensions.Length > 0, nameof(dimensions)); [](start = 33, length = 43)

In an ideal world this check would produce a useful error message, but, I see we don't do it above anyway, so...

Reply:
I don't know. I'd rather not, since I'm not sure what would go in the issue. To identify this specific line would be too petty... but then what should we write in the issue, "many of our usage of `Check` and `CheckParam` have no useful guiding error messages describing what went wrong?" I mean, that's *true*, but I feel like that's not quite capturing the scale of the work, or something.